### ISSUE TITLE
Support "[else]" guard for internal transitions

### ISSUE THREAD

**Issue by iamjuing** [StateSmith/StateSmith#136]
Shall support this scenario w/o a need for using guard check ordering:

![image](https://user-images.githubusercontent.com/125978139/224738729-04da0087-856a-496b-9e88-7066c1f0b2ec.png)

As per discussion: https://github.com/StateSmith/StateSmith/discussions/134

**Comment by adamfk** [comment]
@iamjuing can you download [this example project](https://github.com/StateSmith/StateSmith-examples/tree/main/proposal-else-guard) and try out the implementation? 

Alternatively, you can just copy [this code to your project](https://github.com/StateSmith/StateSmith-examples/blob/main/proposal-else-guard/src/MySm.csx#L13-L37). Might need to copy [some of the usings too](https://github.com/StateSmith/StateSmith-examples/blob/main/proposal-else-guard/src/MySm.csx#L13-L37).

I was thinking specifying an order like `1. EVENT [else]` should be illegal.

**Comment by iamjuing** [comment]
I used the code snippet in my CSX, tried various combinations and it works well.
Yes, any ordering applied to `EVENT [else]` should be definitely considered illegal.

**Comment by iamjuing** [comment]
Found an issue ... with this case:
![image](https://user-images.githubusercontent.com/125978139/229443120-70e832e0-cee5-4825-b3ee-85aa4db462cc.png)

Generated output:
![image](https://user-images.githubusercontent.com/125978139/229443247-d8ef6079-b6f9-48d6-8832-53abe757f212.png)

Problem:
`if (cnt==1)` block does not have a return statement and thus falls to `else` block.

**Comment by iamjuing** [comment]
Could you please make this a priority into an official release?

**Comment by adamfk** [comment]
> Found an issue ... with
> Problem:
> if (cnt==1) block does not have a return statement and thus falls to else block.

That behavior is expected with the current StateSmith event handling. STATE_1 has 3 behaviors:
* EVENT [cnt == 1]
* EVENT [cnt == 2] TransitionTo(STATE_3)
* EVENT [else]  TransitionTo(STATE_2)

`EVENT [cnt == 1]` will prevent an ancestor state from handling `EVENT`, but still allows STATE_1's other `EVENT` behaviors to run. Only a transition will prevent any other behaviors from being checked.

![image](https://user-images.githubusercontent.com/274012/229596069-eac6d697-421f-4d1e-93d2-647bf45a4fad.png)

This is how you can achieve the desired effect:

![image](https://user-images.githubusercontent.com/274012/229596259-447c5447-0e25-46a8-b843-6e17f9bb83df.png)

Although this is clearer:

![diagram 2](https://user-images.githubusercontent.com/274012/229597230-05eccd9c-bc31-47e0-bd9b-d1e4e6c4359a.svg)

(the above svg diagram is editable with draw.io)

Make sense?

**Comment by adamfk** [comment]
> Could you please make this a priority into an official release?

Yep. I'll try to get it out in the next few days.

**Comment by adamfk** [comment]
See plantuml diagram links in tests: https://github.com/StateSmith/StateSmith/blob/bc10435e4756eadf94f85358373d084aa200107e/src/StateSmithTest/SmGraph/ElseGuardProcessorTests.cs

**Comment by iamjuing** [comment]
The solution you suggest is not equivalent as it exits and re-enters the state which is not desirable.
Where does this statement come from: "_Only a transition will prevent any other behaviors from being checked._" ?

The thing is that `EVENT [cnt==1]` is a transition. It is an internal transition (with guard) as per the UML standard.
Therefore , StateSmith shall threat it as a transition and the first `if (cnt==1)` block shall contain the `return` statement to prevent side-effect by actually activating the `EVENT [else]` external transition.

**Comment by adamfk** [comment]
> The solution you suggest is not equivalent as it exits and re-enters the state which is not desirable.

True. I added more examples and my reasoning here: https://github.com/StateSmith/StateSmith/discussions/158 Let's continue the discussion there. These discussions are super helpful to me because they help broaden my perspective so that we can make StateSmith as useful as possible.

I've also added this wiki page to help collect differences with UML:
https://github.com/StateSmith/StateSmith/wiki/UML-Differences
